src/metadata/selectors.js

```diff
@@ -1,225 +1,181 @@
-    const coc = getCategoryOptionComboById(metadata, cocId)
-    return coc.categoryOptions.map((coId) =>
-        getCategoryOptionById(metadata, coId)
-    )
-}
-
-export const getCategoryOptionComboById = (metadata, id) => {
-    return getCategoryOptionCombos(metadata)[id]
-}
-
-export const getCategoriesByCategoryComboId = (metadata, categoryComboId) => {
-    const categoryCombo = getCategoryComboById(metadata, categoryComboId)
-    const categories = categoryCombo.categories.map((id) =>
-        getCategoryById(metadata, id)
-    )
-    return categories
-}
-
-export const getCategoryOptionsByCategoryId = (metadata, categoryId) => {
-    const category = getCategoryById(metadata, categoryId)
-    const categoryOptions = category.categoryOptions.map((optId) =>
-        getCategoryOptionById(metadata, optId)
-    )
-    return categoryOptions
-}
-
-export const getCategoryOptionCombosByCategoryComboId = (
-    metadata,
-    catComboId
-) => {
-    const categoryCombo = getCategoryComboById(metadata, catComboId)
-    const categoryOptionCombos = categoryCombo?.categoryOptionCombos.map(
-        (cocId) => getCategoryOptionComboById(metadata, cocId)
-    )
-    return categoryOptionCombos
-}
+        })
+)((data, dataSetId) => dataSetId)
+
+/**
+ * This selector needs the dataSetId so it can use the selectDataElementsByDataSetId selector,
+ * which will apply the correct categoryCombo overrides.
+ */
+
+export const selectDataElementsBySection = createCachedSelector(
+    (data, _, sectionId) => selectSectionById(data, sectionId),
+    (data, dataSetId) => selectDataElementsByDataSetId(data, dataSetId),
+    (section, dataElements) =>
+        section.dataElements.map((id) =>
+            dataElements.find((de) => de.id === id)
+        )
+)((data, dataSetId, sectionId) => `${dataSetId}:${sectionId}`)
 
 /**
- * Gets the correct ordering of categoryOptionCombos within a category combination
- * @param {*} metadata
- * @param {*} catComboId
- * @returns
+ * Returns an array of objects with dataElements and their associated categoryCombos. They
+ * are grouped by categoryComboId, every time the categoryComboId changes a new group is
+ * created.
  */
-export const getSortedCategoryOptionCombosByCategoryComboId = (
-    metadata,
-    catComboId
-) => {
-    const categories = getCategoriesByCategoryComboId(metadata, catComboId)
-
-    // get options for each category, these should be in sorted order
-    const allOptionsLists = categories.map((cat) => cat.categoryOptions)
-    // compute the  combination of category-catgoryOption- this basically computes the categoryCombinations
-    // result is a matrix of categoryOptions for each column in the categoryCombo
-    const optionCombinationOrder = cartesian(allOptionsLists)
-    return optionCombinationOrder.map((opts) =>
-        getCoCByCategoryOptions(metadata, catComboId, opts)
-    )
-}
+
+export const selectDataElementsByCatComboInOrder = createSelector(
```

Comment:
I think "group" was a pretty integral part of what this function does, and it should not be removed from the name. I think something like `selectGroupedDataElements...` or something would better reflect the intention and result of this function.

I can see why you've renamed the functions, but in general I would be cautious about refactorings like this. A lot of these files are "hot" and are being worked on by others, and renaming them or refactoring stuff creates a lot of merge conflicts to resolve. It's so easy to get into code and think "this could've been done a better way", or "lets follow this naming convention", but I think we should try to resist this urge, especially when there are multiple people working on a project simultaneously.

Reply:
Yeah, I totally understand the rationale, and again not a problem. I think we just have a different perspective on stuff like this. I'm very much a "if it ain't broke, don't fix it kind of guy", and don't really touch code that I don't have to - which is probably just something I've learned through working on the Maintenance app. But there's a strong argument for "fixing broken windows" as well, and do what you think it best - but there's a fine balance. I do think this is particular refactor is justified though.

> For example, everything in selectors.js was exported, but a fair amount of exports were unused.

I really don't see this is a problem. I exported all functions because we will probably need to use the other ones down the line. The selectors are basically an API to the metadata, and I don't see why we should "open"/export this on a case by case basis as they are used in the app.

Reply:
I'll resolve this for now. We can also talk tomorrow and figure out the specifics if there's anything that still needs addressing. Or let me know here in the meantime, good to talk about the coordination of these types of changes.